src/node/plugin.ts

```diff
@@ -1,92 +1,189 @@
-  return originalLoad.apply(this, [request.replace(/^code-server/, path.resolve(__dirname, "../..")), parent, isMain])
+interface Application extends pluginapi.Application {
+  plugin: Plugin
 }
 
 /**
- * Load a plugin and run its activation function.
+ * PluginAPI implements the plugin API described in typings/plugin.d.ts
+ * Please see that file for details.
  */
-const loadPlugin = async (pluginPath: string, app: Express, args: Args): Promise<void> => {
-  try {
-    const plugin: Plugin = require(pluginPath)
-    plugin.activate(app, args)
-
-    const packageJson = require(path.join(pluginPath, "package.json"))
-    logger.debug(
-      "Loaded plugin",
-      field("name", packageJson.name || path.basename(pluginPath)),
-      field("path", pluginPath),
-      field("version", packageJson.version || "n/a"),
-    )
-  } catch (error) {
-    logger.error(error.message)
+export class PluginAPI {
+  private readonly plugins = new Array<Plugin>()
+  private readonly logger: Logger
+
+  public constructor(
+    logger: Logger,
+    /**
+     * These correspond to $CS_PLUGIN_PATH and $CS_PLUGIN respectively.
+     */
+    private readonly csPlugin = "",
+    private readonly csPluginPath = `${path.join(util.paths.data, "plugins")}:/usr/share/code-server/plugins`,
+  ) {
+    this.logger = logger.named("pluginapi")
   }
-}
 
-/**
- * Load all plugins in the specified directory.
- */
-const _loadPlugins = async (pluginDir: string, app: Express, args: Args): Promise<void> => {
-  try {
-    const files = await util.promisify(fs.readdir)(pluginDir, {
-      withFileTypes: true,
-    })
-    await Promise.all(files.map((file) => loadPlugin(path.join(pluginDir, file.name), app, args)))
-  } catch (error) {
-    if (error.code !== "ENOENT") {
-      logger.warn(error.message)
+  /**
+   * applications grabs the full list of applications from
+   * all loaded plugins.
+   */
+  public async applications(): Promise<Application[]> {
+    const apps = new Array<Application>()
+    for (const p of this.plugins) {
+      const pluginApps = await p.applications()
+
+      // Add plugin key to each app.
+      apps.push(
+        ...pluginApps.map((app) => {
+          return { ...app, plugin: p }
+        }),
+      )
     }
+    return apps
   }
-}
 
-/**
- * Load all plugins from the `plugins` directory, directories specified by
- * `CS_PLUGIN_PATH` (colon-separated), and individual plugins specified by
- * `CS_PLUGIN` (also colon-separated).
- */
-export const loadPlugins = async (app: Express, args: Args): Promise<void> => {
-  const pluginPath = process.env.CS_PLUGIN_PATH || `${path.join(paths.data, "plugins")}:/usr/share/code-server/plugins`
-  const plugin = process.env.CS_PLUGIN || ""
-  await Promise.all([
+  /**
+   * mount mounts all plugin routers onto r.
+   */
+  public mount(r: express.Router): void {
+    for (const p of this.plugins) {
+      r.use(`/${p.name}`, p.router())
```

Comment:
This is more of a minor concern, but the routes are going to be really ugly if these plugins come from npm. They'll all be something like `/code-server-plugin-jupyter/ide` or something to that effect.

Reply:
That's fair. Perhaps I should move name back into the Plugin interface then.

Reply:
Actually moving this into the plugin interface. I think only the version should be taken from `package.json`.